### PR TITLE
fix(cdk/overlay): detach overlay when portal is destroyed from the outside

### LINKED ISSUE #25163: bug(MatDialog): Providing ViewContainerRef to MatDialog and removing the providing component leaves overlay in place
### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

_No response_

### Description

There seems to be a side effect to this "issue": https://github.com/angular/components/issues/25073
When providing viewContainerRef to the dialog the dialog/overlay cannot be closed/removed if the owning component is removed before the dialog is closed.

### Reproduction

https://stackblitz.com/edit/angular-14-provider-bug-wfnbti?file=src/app/dialog/dialog.component.ts

Steps to reproduce:
1. Open application
2. Wait for dialog to close

### Expected Behavior

I expect the dialog to either still be open and that it can be closed later without being left with a empty dialog and a backdrop that can't be closed or that the backdrop is also removed if it's the last or only dialog.

### Actual Behavior

In Angular 13.x the dialog is not automatically removed when the component owning the viewContainerRef is removed and when I close the dialog later on I'm left with a tiny empty dialog and the backdrop/overlay.

In Angular 14.x the dialog is automatically closed when the owning component is removed, but I'm left with the overlay/backdrop and no way to close it.

### Environment

Angular CLI: 14.0.1
Node: 16.14.0
Package Manager: npm 8.3.1
OS: win32 x64

Angular: 14.0.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, material, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1400.1
@angular-devkit/build-optimizer   0.1302.1
@angular-devkit/core              14.0.1
@angular-devkit/schematics        14.0.1
@ngtools/webpack                  14.0.1
@schematics/angular               14.0.1
ng-packagr                        14.0.2
rxjs                              7.5.5
typescript                        4.7.3
webpack                           5.73.0

### PR DESCRIPTION
**Pull request by crisbeto**
Fixes that some of the overlay elements were left behind when the portal is destroyed without going through the overlay API. We need to handle this case, because in many cases users don't have access to the `OverlayRef` (e.g. `MatDialog`) so that they can call `dispose` themselves.

Fixes #25163.

### CODE REVIEW

File: src/cdk/overlay/overlay-ref.ts [hunk 1]
```diff
@@ -158,6 +158,26 @@ export class OverlayRef implements PortalOutlet, OverlayReference {
...
+      // we can finish the disposal process. The exception is if the user passes in a custom
+      // `ViewContainerRef` that isn't destroyed through the overlay API. Note that we use
+      // `detach` here instead of `dispose`, because we don't know if the user intends to
+      // reattach the overlay at a later point. It also has the advantage of waiting for animations.
+      attachResult.onDestroy(() => {
+        if (this.hasAttached()) {
+          // We have to wrap the `detach` call in `onStable`, because detaching immediately prevents
+          // other destroy hooks from running. This is likely a framework bug similar to
+          // https://github.com/angular/angular/issues/46119
+          this._ngZone.onStable.pipe(take(1)).subscribe(() => this.detach());
```
Comment (jelbourn) [inline src/cdk/overlay/overlay-ref.ts:176]: Should this also happen outside the zone?

Comment (jelbourn) [reply src/cdk/overlay/overlay-ref.ts:176]: If it is causing another change detection (not 100% sure it would be, but I suspect it would), it could lead to apps depending on the extra change detection in tests

**Comment by crisbeto** [comment]
The feedback has been addressed.

**Review by andrewseguin — APPROVED** [review]
Curious to see if this requires any test timing updates internally
